staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go

```diff
@@ -0,0 +1,259 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package buffered implements the audit.Backend interface with buffered channel.
+package buffered
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/client-go/util/flowcontrol"
+)
+
+const (
+	// Default configuration values for ModeBatch.
+	//
+	// TODO(ericchiang): Make these value configurable. Maybe through a
+	// kubeconfig extension?
+	defaultBatchBufferSize = 10000            // Buffer up to 10000 events before starting discarding.
+	defaultBatchMaxSize    = 400              // Only consume up to 400 events at a time.
+	defaultBatchMaxWait    = 30 * time.Second // Consume events at least twice a minute.
+
+	defaultBatchThrottleQPS   = 10 // Limit the consume rate by 10 QPS.
+	defaultBatchThrottleBurst = 15 // Allow up to 15 QPS burst.
+)
+
+// The plugin name reported in error metrics.
+const pluginName = "buffered"
+
+// NewBackend returns an audit buffered backend.
+func NewBackend(backend audit.Backend) audit.Backend {
+	return &bufferedBackend{
+		backend:      backend,
+		buffer:       make(chan *auditinternal.Event, defaultBatchBufferSize),
+		maxBatchSize: defaultBatchMaxSize,
+		maxBatchWait: defaultBatchMaxWait,
+		shutdownCh:   make(chan struct{}),
+		throttle:     flowcontrol.NewTokenBucketRateLimiter(defaultBatchThrottleQPS, defaultBatchThrottleBurst),
+	}
+}
+
+type bufferedBackend struct {
+	backend audit.Backend
+	// Channel to buffer events in memory before consuming them with backend worker routines.
```

Comment:
I think mentioning `working routines` here is confusing

Additionally in all following comments I think the word `consume` is confusing, what about using `sending/passing/feeding to an underlying backend` instead?

Reply:
How about `Channel to buffer events before underlying backend processing.`